crates/matrix-sdk-crypto-nodejs/src/ts/OlmMachine.ts

```diff
@@ -0,0 +1,318 @@
+    signatures: Signatures;
+}
+
+export interface OTKCounts {
+    one_time_key_counts: {
+        [algorithm: string]: number;
+    };
+    fallback_keys?: {
+        [algorithm: string]: FallbackKey;
+    };
+    ["org.matrix.msc2732.fallback_keys"]?: {
+        [algorithm: string]: FallbackKey;
+    };
+}
+
+export interface KeyClaim {
+    [userId: string]: {
+        [deviceId: string]: string;
+    };
+}
+
+export interface KeyClaimResponse {
+    one_time_keys: {
+        [userId: string]: {
+            [deviceId: string]: {
+                [algorithm: string]: string | {
+                    key: string;
+                    signatures: Signatures;
+                };
+            };
+        };
+    };
+    failures?: {
+        [serverName: string]: Record<string, any>;
+    };
+}
+
+export interface KeyQueryResults {
+    device_keys: {
+        [userId: string]: {
+            [deviceId: string]: {
+                algorithms: string[];
+                device_id: string;
+                keys: {
+                    [keyId: string]: string;
+                };
+                signatures: Signatures;
+                unsigned?: Record<string, any>;
+                user_id: string;
+            };
+        };
+    };
+    master_keys?: {
+        [userId: string]: {
+            keys: {
+                [keyId: string]: string;
+            };
+            usage: string[];
+            user_id: string;
+        };
+    };
+    self_signing_keys?: {
+        [user_id: string]: {
+            keys: {
+                [keyId: string]: string;
+            };
+            signatures: Signatures;
+            usage: string[];
+            user_id: string;
+        };
+    };
+    user_signing_keys?: {
+        [user_id: string]: {
+            keys: {
+                [keyId: string]: string;
+            };
+            signatures: Signatures;
+            usage: string[];
+            user_id: string;
+        };
+    };
+    failures?: {
+        [serverName: string]: Record<string, any>;
+    };
+}
+
+export interface ToDeviceMessages {
+    [userId: string]: {
+        [deviceId: string]: any;
+    };
+}
+
+export interface OlmEngine {
+    uploadOneTimeKeys(body: {device_keys?: DeviceKeys, one_time_keys?: GenericKeys}): Promise<OTKCounts>;
+    queryOneTimeKeys(userIds: string[]): Promise<KeyQueryResults>;
+    claimOneTimeKeys(claim: KeyClaim): Promise<KeyClaimResponse>;
+    sendToDevices(eventType: string, messages: ToDeviceMessages): Promise<void>;
+    getEffectiveJoinedUsersInRoom(roomId: string): Promise<string[]>;
+}
+
+// Dev note: Due to complexities in the types structure, we can't reliably pull in the types for the
+// napi-module. Be careful with your implementation!
+
+export class OlmMachine {
+    private machine: any; // SledBackedOlmMachine
+
+    private constructor(private _userId: string, private _deviceId: string, private engine: OlmEngine) {
+    }
+
+    private makeSled(dir: string): OlmMachine {
```

Comment:
Is this useful if you have `withSledBackend()`?

Reply:
it's the way we do factories in JS, yea. `withSledBackend` is static so the values don't transfer to the inner instance.